analyzers/src/SonarAnalyzer.Common/Rules/DoNotCheckZeroSizeCollectionBase.cs

```diff
@@ -19,124 +19,134 @@
-                LessThanExpression);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
 
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
                 c =>
                 {
-                    var binaryExpression = (TBinaryExpressionSyntax)c.Node;
-                    CheckCondition(c, c.Node, GetRightNode(binaryExpression), GetLeftNode(binaryExpression));
+                    var binary = c.Node;
+                    var binaryLeft = Language.Syntax.BinaryExpressionLeft(binary);
+                    var binaryRight = Language.Syntax.BinaryExpressionRight(binary);
+
+                    if (Language.ExpressionNumericConverter.TryGetConstantIntValue(c.SemanticModel, binaryLeft, out var left))
+                    {
+                        CheckExpression(c, binary, binaryRight, left, Language.Syntax.ComparisonKind(binary).Mirror());
+                    }
+                    else if (Language.ExpressionNumericConverter.TryGetConstantIntValue(c.SemanticModel, binaryRight, out var right))
+                    {
+                        CheckExpression(c, binary, binaryLeft, right, Language.Syntax.ComparisonKind(binary));
+                    }
                 },
-                LessThanOrEqualExpression,
-                GreaterThanExpression);
-        }
+                Language.SyntaxKind.ComparisonKinds);
 
-        protected void CheckCondition(SyntaxNodeAnalysisContext context, SyntaxNode issueLocation, TExpressionSyntax expressionValueNode, TExpressionSyntax constantValueNode)
+        protected void CheckExpression(SyntaxNodeAnalysisContext context, SyntaxNode issue, SyntaxNode expression, int constant, ComparisonKind comparison)
         {
-            if (IsZeroOrNegativeConstantValue(context, constantValueNode)
-                && GetSymbol(context, expressionValueNode) is { } symbol
+            var result = comparison.Compare(constant);
+            if (result.IsInvalid()
+                && CollectionSizeSymbol(Language.Syntax.RemoveConditionalAcesss(expression), context.SemanticModel) is { } symbol
                 && GetDeclaringTypeName(symbol) is { } symbolType)
             {
-                context.ReportIssue(Diagnostic.Create(SupportedDiagnostics[0], issueLocation.GetLocation(), symbol.Name, symbolType));
+                context.ReportIssue(Diagnostic.Create(rule, issue.GetLocation(), symbol.Name, symbolType, result == CountComparisonResult.AlwaysTrue));
             }
         }
 
+        private ISymbol CollectionSizeSymbol(SyntaxNode expression, SemanticModel semanticModel) =>
+            HasCandidateName(Language.Syntax.NodeIdentifier(expression)?.ValueText)
+            && semanticModel.GetSymbolInfo(expression).Symbol is ISymbol symbol
+            && IsCollecionSizeCheck(symbol)
+            ? symbol : null;
```

Comment:
I don't like the fact that `IsCollecionSizeCheck` and `GetDeclaringTypeName` do the same thing. And `GetDeclaringTypeName` has hidden dependency on `IsCollecionSizeCheck` making the `return null` unreachable.
Let's:
* Remove `IsCollecionSizeCheck` completely
* Rename `GetDeclaringTypeName` to `CollecionSizeTypeName`
* Remove `CollectionSizeSymbol` and move it's content to `CheckExpression` like this:
```
            var result = comparison.Compare(constant);
            expression = Language.Syntax.RemoveConditionalAcesss(expression);
            if (result.IsInvalid()
                && HasCandidateName(Language.Syntax.NodeIdentifier(expression)?.ValueText)
                && context.SemanticModel.GetSymbolInfo(expression).Symbol is ISymbol symbol
                && CollecionSizeTypeName(symbol) is { } symbolType)
```